swift_browser_ui/common/vault_client.py

```diff
@@ -0,0 +1,295 @@
+"""Vault c4ghtransit client"""
+import asyncio
+import logging
+import os
+import time
+from base64 import standard_b64encode
+from dataclasses import dataclass
+from typing import Any, Dict, List, Optional, Union
+
+from aiohttp import ClientSession, ClientTimeout
+from aiohttp.web import HTTPError, HTTPGatewayTimeout, HTTPInternalServerError
+from yarl import URL
+
+LOGGER = logging.getLogger("swift_browser_ui.common.vault_client")
+LOGGER.setLevel(os.environ.get("LOG_LEVEL", "INFO"))
+
+
+class VaultServerError(HTTPError):
+    """Service server errors should produce a 502 Bad Gateway response."""
+
+    status_code = 502
+
+
+class VaultClientError(HTTPError):
+    """Service client errors should be raised unmodified."""
+
+    def __init__(
+        self,
+        status_code: int,
+        **kwargs: Any,
+    ) -> None:
+        """Class to raise for http client errors.
+
+        HTTPError doesn't have a setter for status_code, so this allows setting it.
+
+        :param status_code: Set the status code here
+        """
+        self.status_code = status_code
+        HTTPError.__init__(self, **kwargs)
+
+
+@dataclass
+class VaultError:
+    """Unpack a vault error into an object."""
+
+    errors: List[str]
+
+
+def _make_exception(reason: str, status: int) -> HTTPError:
+    """Create a Client or Server exception, according to status code.
+
+    https://developer.hashicorp.com/vault/api-docs#http-status-codes
+
+    :param reason: Error message
+    :param status: HTTP status code
+    :returns VaultServerError or VaultClientError.
+    """
+    if status < 400:
+        LOGGER.error("HTTP status code must be an error code, >400 received %s.", status)
+        return HTTPInternalServerError(
+            reason="Server encountered an unexpected situation."
+        )
+    reason = f"Vault error: {reason}"
+    if status >= 500:
+        return VaultServerError(text=reason, reason=reason)
+    return VaultClientError(text=reason, reason=reason, status_code=status)
+
+
+def _process_error(vault_error: VaultError) -> str:
+    """Return formatted error message.
+
+    https://developer.hashicorp.com/vault/api-docs#error-response
+
+    :param vault_error: Vault error dictionary object
+    :returns: formatted error message as a string.
+    """
+    return vault_error.errors[0]
+
+
+class VaultClient:
+    """AioHttp client that interacts with a Vault server."""
+
+    _vault_token = str()
+    service = "SD-Connect"
+
+    def __init__(self, http_client: ClientSession):
+        """Reuse an existing http client, and load environment variables.
+
+        :param http_client: aiohttp.ClientSession instance
+        """
+        self._client = http_client
+        self._base_url = URL(os.environ.get("VAULT_URL", "http://localhost:8200")) / "v1"
```

Comment:
```suggestion
        self._base_url = URL(os.environ.get("VAULT_URL", "http://localhost:8200/v1"))
```
I think `v1` should be part of the configurable address even if it is unlikely to change in the near future

Reply:
Seconding this, no need to have the `v1` part hardcoded in the code, even though I don't expect it to change. Other API URLs are configured in a similar manner, like `/v3` in Openstack Keystone URLs.

Reply:
ok